### DISCUSSION TITLE
Writing tests in Rust for a python extension

### DISCUSSION THREAD

**Discussion by Adoni5** [PyO3/pyo3#2733] [Questions] [answered]
Hi,

Apologies for what may be a reposted question. although I had a cursory look and could not find anything. I am sure that when I was reading the docs.rs, I saw an example of how to write rust tests for your python extension, using _something_ loosely along the lines of 

```rust
     Python::with_GIL(py || {
         _example test_
    })
```

However I cannot for the life of me find it again. I was wondering if there was any standard way of writing these tests?

Thanks for the brilliant crate!
Rory

**Comment by davidhewitt** [comment, marked as answer]
Hi, sorry for the slow reply.

You can indeed write tests from Rust, you should make sure to disable the `extension-module` feature and instead enable the `auto-initialize` feature while running tests. 

For some examples, see our tests we have internally for string conversions: https://github.com/PyO3/pyo3/blob/23fcdf950b94df5136b81f53e55f22752c84dbc7/src/conversions/std/string.rs#L149-L164

**Reply by Adoni5** [reply]
Hi David,

Thanks for the reply and the pointer to some examples! That should do it.
